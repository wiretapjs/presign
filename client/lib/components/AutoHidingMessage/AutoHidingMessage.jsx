import React, { Component } from 'react'
import PropTypes from 'prop-types'

class AutoHidingMessage extends Component {
  constructor (props) {
    super(props)
    const ahd = this.props.autoHideDuration
    this.state = {
      display: this.props.display,
      // assumes values less than 1000 are seconds rather than ms
      autoHideDuration: ahd < 1000 ? ahd * 1000 : ahd,
    }
  }

  render () {
    if (this.state.display) {
      return (
        <div style={this.props.divStyle}>
          <span style={this.props.textStyle}>{this.props.message}</span>
        </div>
      )
    }
    return null
  }

  componentDidMount () {
    this.setAutoHide()
  }

  componentWillReceiveProps (newProps) {
    this.setState({
      display: newProps.display,
      message: newProps.message,
      textStyle: newProps.textStyle,
      divStyle: newProps.divStyle,
    }, () => {
      this.setAutoHide()
    })
  }

  setAutoHide () {
    if (this.state.display && this.state.autoHideDuration) {
      setTimeout(() => {
        if (this.state.display) {
          this.setState({
            display: false,
          })
        }
      }, this.state.autoHideDuration)
    }
  }
}

AutoHidingMessage.propTypes = {
  display: PropTypes.bool,
  message: PropTypes.string,
  autoHideDuration: PropTypes.number,
  textStyle: PropTypes.object,
  divStyle: PropTypes.object,
}

AutoHidingMessage.defaultProps = {
  display: false,
  message: '',
  autoHideDuration: null,
  textStyle: null,
  divStyle: null,
}

export default AutoHidingMessage
