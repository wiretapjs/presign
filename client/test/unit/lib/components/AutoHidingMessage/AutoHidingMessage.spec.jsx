import React from 'react'
import { shallow, mount } from 'enzyme'
import AutoHidingMessage from 'lib/components/AutoHidingMessage/AutoHidingMessage'

describe('AutoHidingMessage', () => {
  it('should render correctly', () => {
    const ahmsg = shallow(<AutoHidingMessage display />)
    expect(ahmsg).toMatchSnapshot()
  })

  it('should not render anything when display is false', () => {
    const ahmsgWrapper = mount(<AutoHidingMessage display={false} />)
    const ahmInstance = ahmsgWrapper.instance()
    expect(ahmInstance.state.display).toBe(false)
  })

  it('should render when display is true', () => {
    const ahmsgWrapper = mount(<AutoHidingMessage display autoHideDuration={1} />)
    const ahmInstance = ahmsgWrapper.instance()
    expect(ahmInstance.state.display).toBe(true)
  })

  describe('Receiving props', () => {
    it('should set display from new props', async () => {
      const autoHidingMessageWrapper = shallow(<AutoHidingMessage />)
      const autoHidingMessageInstance = autoHidingMessageWrapper.instance()
      await autoHidingMessageInstance.componentWillReceiveProps({ display: true })
      expect(autoHidingMessageInstance.state.display).toEqual(true)
    })

    it('should set message from new props', async () => {
      const autoHidingMessageWrapper = shallow(<AutoHidingMessage />)
      const autoHidingMessageInstance = autoHidingMessageWrapper.instance()
      await autoHidingMessageInstance.componentWillReceiveProps({ message: 'test message' })
      expect(autoHidingMessageInstance.state.message).toEqual('test message')
    })

    it('should set textStyle from new props', async () => {
      const autoHidingMessageWrapper = shallow(<AutoHidingMessage />)
      const autoHidingMessageInstance = autoHidingMessageWrapper.instance()
      await autoHidingMessageInstance.componentWillReceiveProps({ textStyle: { color: 'red' } })
      expect(autoHidingMessageInstance.state.textStyle).toEqual({ color: 'red' })
    })

    it('should set textStyle from new props', async () => {
      const autoHidingMessageWrapper = shallow(<AutoHidingMessage />)
      const autoHidingMessageInstance = autoHidingMessageWrapper.instance()
      await autoHidingMessageInstance.componentWillReceiveProps({ divStyle: { paddingTop: 5, maxWidth: '375px' } })
      expect(autoHidingMessageInstance.state.divStyle).toEqual({ paddingTop: 5, maxWidth: '375px' })
    })
  })

  describe('Autohiding', () => {
    it('should hide when it is displayed and an autohide duration is set', async () => {
      jest.useFakeTimers()

      const autoHidingMessageWrapper = shallow(<AutoHidingMessage />)
      const autoHidingMessageInstance = autoHidingMessageWrapper.instance()
      autoHidingMessageInstance.setState({ display: true, autoHideDuration: 1 })
      await autoHidingMessageInstance.setAutoHide()

      setTimeout(() => {
        expect(autoHidingMessageInstance.state.display).toEqual(false)
      }, autoHidingMessageInstance.state.autoHideDuration)

      jest.runAllTimers()
    })

    it('should not hide when it is displayed and an autohide duration is set, and then hidden', async () => {
      jest.useFakeTimers()

      const autoHidingMessageWrapper = shallow(<AutoHidingMessage />)
      const autoHidingMessageInstance = autoHidingMessageWrapper.instance()
      autoHidingMessageInstance.setState({ display: true, autoHideDuration: 1 })
      await autoHidingMessageInstance.setAutoHide()
      autoHidingMessageInstance.setState({ display: false })

      setTimeout(() => {
        expect(autoHidingMessageInstance.state.display).toEqual(false)
      }, autoHidingMessageInstance.state.autoHideDuration)

      jest.runAllTimers()
    })

    it('should not hide when it is not displayed and an autohide duration is set', async () => {
      const autoHidingMessageWrapper = shallow(<AutoHidingMessage />)
      const autoHidingMessageInstance = autoHidingMessageWrapper.instance()
      autoHidingMessageInstance.setState({ display: false, autoHideDuration: undefined })
      await autoHidingMessageInstance.setAutoHide()
      expect(autoHidingMessageInstance.state.display).toEqual(false)
    })

    it('should not hide when it is not displayed and an autohide duration is set', async () => {
      const autoHidingMessageWrapper = shallow(<AutoHidingMessage />)
      const autoHidingMessageInstance = autoHidingMessageWrapper.instance()
      autoHidingMessageInstance.setState({ display: false, autoHideDuration: undefined })
      await autoHidingMessageInstance.setAutoHide()
      expect(autoHidingMessageInstance.state.display).toEqual(false)
    })

    it('should not hide when it is not displayed and an autohide duration is not set', async () => {
      const autoHidingMessageWrapper = shallow(<AutoHidingMessage />)
      const autoHidingMessageInstance = autoHidingMessageWrapper.instance()
      autoHidingMessageInstance.setState({ display: false, autoHideDuration: undefined })
      await autoHidingMessageInstance.setAutoHide()
      expect(autoHidingMessageInstance.state.display).toEqual(false)
    })

    it('should set autohide duration as seconds when the value is less than 1000', () => {
      const autoHidingMessageWrapper = shallow(<AutoHidingMessage autoHideDuration={1} />)
      const autoHidingMessageInstance = autoHidingMessageWrapper.instance()
      expect(autoHidingMessageInstance.state.autoHideDuration).toEqual(1000)
    })

    it('should set autohide duration as milliseconds when the value is greater or equal to 1000', () => {
      const autoHidingMessageWrapper = shallow(<AutoHidingMessage autoHideDuration={1000} />)
      const autoHidingMessageInstance = autoHidingMessageWrapper.instance()
      expect(autoHidingMessageInstance.state.autoHideDuration).toEqual(1000)
    })
  })
})
