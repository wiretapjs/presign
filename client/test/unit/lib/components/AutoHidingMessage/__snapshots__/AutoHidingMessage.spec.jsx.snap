// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoHidingMessage should render correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    style={null}
  >
    <span
      style={null}
    >
      
    </span>
  </div>,
  "nodes": Array [
    <div
      style={null}
    >
      <span
        style={null}
      >
        
      </span>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <AutoHidingMessage
        autoHideDuration={null}
        display={true}
        divStyle={null}
        message=""
        textStyle={null}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": AutoHidingMessage {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "autoHideDuration": null,
          "display": true,
          "divStyle": null,
          "message": "",
          "textStyle": null,
        },
        "refs": Object {},
        "state": Object {
          "autoHideDuration": 0,
          "display": true,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          style={null}
        >
          <span
            style={null}
          >
            
          </span>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          style={null}
        >
          <span
            style={null}
          >
            
          </span>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <AutoHidingMessage
    autoHideDuration={null}
    display={true}
    divStyle={null}
    message=""
    textStyle={null}
  />,
}
`;
